server/src/main/java/org/apache/druid/client/MetadataSegmentView.java

```diff
@@ -0,0 +1,200 @@
+    this.jsonMapper = jsonMapper;
+    this.responseHandler = responseHandler;
+    this.segmentWatcherConfig = segmentWatcherConfig;
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    scheduledExec = Execs.scheduledSingleThreaded("MetadataSegmentView-Cache--%d");
+    scheduledExec.schedule(new PollTask(), 0, TimeUnit.MILLISECONDS);
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    scheduledExec.shutdownNow();
+    scheduledExec = null;
+  }
+
+  private void poll()
+  {
+    log.info("polling published segments from coordinator");
+    //get authorized published segments from coordinator
+    final JsonParserIterator<DataSegment> metadataSegments = getMetadataSegments(
+        coordinatorDruidLeaderClient,
+        jsonMapper,
+        responseHandler,
+        segmentWatcherConfig.getWatchedDataSources()
+    );
+
+    final DateTime timestamp = DateTimes.nowUtc();
+    while (metadataSegments.hasNext()) {
+      final DataSegment interned = DataSegmentInterner.intern(metadataSegments.next());
+      // timestamp is used to filter deleted segments
+      publishedSegments.put(interned, timestamp);
+    }
+    // filter the segments from cache whose timestamp is not equal to latest timestamp stored,
+    // since the presence of a segment with an earlier timestamp indicates that
+    // "that" segment is not returned by coordinator in latest poll, so it's
+    // likely deleted and therefore we remove it from publishedSegments
+    final Set<DateTime> toBeRemovedSegments = publishedSegments.values()
+                                                         .stream()
+                                                         .filter(v -> v != timestamp)
+                                                         .collect(Collectors.toSet());
+    publishedSegments.values().removeAll(toBeRemovedSegments);
+
+    if (segmentWatcherConfig.getWatchedDataSources() != null) {
+      log.debug(
+          "filtering datasources[%s] in published segments based on broker's watchedDataSources",
+          segmentWatcherConfig.getWatchedDataSources()
+      );
+      publishedSegments.keySet()
+                       .removeIf(key -> !segmentWatcherConfig.getWatchedDataSources().contains(key.getDataSource()));
+    }
+  }
+
+  public Iterator<DataSegment> getPublishedSegments()
+  {
+    return publishedSegments.keySet().iterator();
+  }
+
+  // Note that coordinator must be up to get segments
+  private static JsonParserIterator<DataSegment> getMetadataSegments(
+      DruidLeaderClient coordinatorClient,
+      ObjectMapper jsonMapper,
+      BytesAccumulatingResponseHandler responseHandler,
+      Set<String> watchedDataSources
+  )
+  {
+    String query = "/druid/coordinator/v1/metadata/segments";
+    if (watchedDataSources != null && !watchedDataSources.isEmpty()) {
+      final StringBuilder sb = new StringBuilder();
+      for (String ds : watchedDataSources) {
+        sb.append("datasources=" + ds + "&");
```

Comment:
```sb.append("datasources=").append(ds).append("&")```

Reply:
You also need to delete the last `&`.

Reply:
Ok. Please check https://github.com/apache/incubator-druid/pull/6901/files#r252479790.

Reply:
oh, missed that one